pmd-java/src/main/java/net/sourceforge/pmd/lang/java/ast/ASTMethodDeclaration.java

```diff
@@ -108,4 +111,25 @@ public ASTNameList getThrows() {
         }
         return null;
     }
+
+    @Override
+    public QualifiedName getQualifiedName() {
+        if (qualifiedName != null) {
+            return qualifiedName;
+        }
+
+        QualifiedName parent = this.getFirstParentOfType(ASTClassOrInterfaceDeclaration.class).getQualifiedName();
+
+        ASTFormalParameters params = this.getFirstDescendantOfType(ASTFormalParameters.class);
+        int numParams = params.jjtGetNumChildren();
+        String[] types = new String[numParams];
+
+        for (int i = 0; i < numParams; i++) {
+            types[i] = params.jjtGetChild(i).getFirstDescendantOfType(ASTType.class).getTypeImage();
```

Comment:
That's indeed much cleaner.

I've refactored those building procedures so that they take the node object as parameter. That encapsulates more the implementation of QualifiedName and removes the need for setters, which makes QualifiedNames nicely immutable.